File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* validSequence(char* word1, char* word2, int* returnSize) {
    int n = strlen(word1);
    int m = strlen(word2);

    int* suffix_match = (int*)malloc((m + 1) * sizeof(int));
    for(int i = 0; i <= m ; i++) suffix_match[i] = -1;
    suffix_match[m] = n;

    int i = n - 1;
    for(int j = m - 1; j >= 0; j--){
        while(i >= 0 && word1[i] != word2[j]){
            i--;
        }
        if(i >= 0){
            suffix_match[j] = i;
            i--;
        }else{
            break;
        }
    }
    
    int* result = (int*)malloc(m * sizeof(int));
    int res = 0;
    i = 0;
    bool changed = false;

    for(int j = 0; j < m; j++){
        if(i < n && word1[i] == word2[j]){
            result[res++] = i;
            i++;
        }else{
            if(!changed && i < n && suffix_match[j + 1] > i){
                result[res++] = i;
                i++;
                changed = true;
            }else{
                while(i < n && word1[i] != word2[j]){
                    i++;
                }
                if(i < n){
                    result[res++] = i;
                    i++;
                }else{
                    free(suffix_match);
                    free(result);
                    *returnSize = 0;
                    return NULL;
                }
            }
        }
    }
    free(suffix_match);
    if(res == m){
        *returnSize = m;
        return result;
    }else{
        free(result);
        *returnSize = 0;
        return NULL;
    }
}

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> List[int]:
        n = len(word1)
        m = len(word2)
        suffix_match = [-1] * (m + 1)
        suffix_match[m] = n
        i = n - 1
        for j in range(m - 1 , -1 , -1):
            while i >= 0 and word1[i] != word2[j]:
                i -= 1
            if i >= 0:
                suffix_match[j] = i
                i -= 1
            else:
                break
        result = []
        i = 0
        changed = False
        for j in range(m):
            if i < n and word1[i] == word2[j]:
                result.append(i)
                i += 1
            else:
                if not changed and i < n and suffix_match[j + 1] > i:
                    result.append(i)
                    i += 1
                    changed = True
                else:
                    while i < n and word1[i] != word2[j]:
                        i += 1
                    if i < n:
                        result.append(i)
                        i += 1
                    else:
                        return []
        return result if len(result) == m else []